### DISCUSSION TITLE
k3s rootless and cgroups v2

### DISCUSSION THREAD

**Discussion by FlorianLudwig** [k3s-io/k3s#3561] [Q&A] [answered]
Starting k3s in rootless mode fails with `failed to find cpuset cgroup (v2)`. My system does have cgroups v2 enabled though (e.g. `/sys/fs/cgroup/cgroup.controllers` is present).

Welcoming any pointers how to debug what the root cause is

**Comment by brandond** [comment, marked as answer]
Which version of k3s are you using? Do you have pure v2 cgroups, or are you in hybrid mode where some of them are still split out?

**Reply by FlorianLudwig** [reply]
```
k3s --version
k3s version v1.21.2+k3s1 (5a67e8dc)
go version go1.16.4
```

On fedora 34.

regarding v2 cgroups - I don't think I have the hybrid on. I didn't know there is a hybrid mode - not exactly sure how to determine if it is on.

**Reply by AkihiroSuda** [reply]
Please make sure:
- k3s is running as a systemd service https://github.com/k3s-io/k3s/blob/master/k3s-rootless.service
- systemd delegation is enabled  https://rootlesscontaine.rs/getting-started/common/cgroup2/

**Reply by FlorianLudwig** [reply]
Hi @AkihiroSuda,

cgroup delegation:
```
$ cat /sys/fs/cgroup/user.slice/user-$(id -u).slice/user@$(id -u).service/cgroup.controllers
cpu io memory pids
```

So indeed `cpuset` is missing. I followed the instructions [there](https://rootlesscontaine.rs/getting-started/common/cgroup2/#enabling-cpu-cpuset-and-io-delegation) and now I do get:

```
$ cat /sys/fs/cgroup/user.slice/user-$(id -u).slice/user@$(id -u).service/cgroup.controllers
cpuset cpu io memory pids
```

Now I do not run into that error anymore but a different one. I am looking into that and will open a new discussion - if needed - as this one is resolved now :) Thanks!
